[FAB-17831] Use generic constant/var names for ledger data format

wenjianqiao opened the pull request:
Signed-off-by: Wenjian Qiao <wenjianq@gmail.com>

#### Type of change
- Improvement (improvement to code, performance, etc)

#### Description
Currently Version1x and Version20 are defined in common/ledger/dataformat.go.
They should be renamed to more generic names such as PreviousVersion and CurrentVersion.
Note that these versions are used for the data format in ledger databases. They would change only when the data format is changed. 

#### Related issues
https://jira.hyperledger.org/browse/FAB-17831

srderson commented:
Why does rbac.yaml contain a complete copy of core.yaml? My concern is that we will not be able to keep these in sync.

Also, is it possible to include a README with this chaincode describing its functionality? I'd preferably like to do that for all example chaincodes.

adecaro commented:
Hi @srderson, this chaincode is part of a joint effort with @jeffgarratt to write behave tests for the access control mechanism supported by the fabric.

A copy of core.yaml is there for test purposes. It will not stay there forever. In fact, once the client sdk will be ready I will write an app to show how to invoke the chaincode leveraging the client sdk drectly. Actually, having a test that shows how to invoke such a chaincode will help us driving the design of the client sdk to ensure that all the required API are exposed as expected. What do you think?

For the README, I will write it soon :)

srderson commented:
@adecaro @jeffgarratt So it looks like this chaincode is deployed in the `rbac_test.go` test case? Is it possible to use a behave test instead? The unit tests where chaincodes are deployed are generally fragile. Ideally I think they should all be moved over to behave tests.

adecaro commented:
@srderson, Indeed, rbac_test simulates a minimal fabric environment, deploys the chaincode and invokes it in order to test its behavior. Actually, the final goal is to have behave tests. I'm just concerned that without the client sdk, it would be difficult to run successfully any behave test because the CLI (nor the REST API) does not provide the required API to drive the chaincode.

Anyway, I will try to understand if we can have something cleaner without waiting the client sdk before merging :)

jeffgarratt commented:
@adecaro @srderson The behave test could be implemented as part of the effort I am working on now.  Either #1139 or #961.

binhn commented:
@srderson  should we wait for the behave tests to merge or just merge it since it is a example chaincode?

srderson commented:
@binhn I'd like to avoid merging `rbac.yaml` and `rbac_test.go`. The current effort is to remove these types of tests, not add more. Also, I'd really like to see a README with all example chaincodes, at least documenting their purpose and API.

adecaro commented:
@binhn @srderson  I also think we can wait the behave tests to merge. I'm working with @jeffgarratt to have them ready asap.

christo4ferris commented:
@adecaro could you please fix the license in these new files? It should be the text that I have in PR #1288 thanks

jeffgarratt commented:
@adecaro @srderson I need to touch base with @adecaro and work this workflow.  Have come to a block point based upon current flow (which is already a bit precarious).

adecaro commented:
Hi @christo4ferris, I have updated the license and also added a readme file to describe the chaincode. 

@christo4ferris, @jeffgarratt, @srderson I can also add a demo app like I did for the asset management chaincode to show how to invoke it. What do you think?

BiscuitBomber commented on examples/chaincode/go/rbac/README.md:
```diff
@@ -0,0 +1,82 @@
+1. Alice is the deployer of the chaincode;
+2. Alice wants to assign the administrator role to Bob;
+3. Alice obtains, via an out-of-band channel, a TCert of Bob, let us call this certificate *BobCert*;
+4. Alice constructs a deploy transaction, as described in *application-ACL.md*,  setting the transaction metadata to *DER(BobCert)*.
```
Please can we make this a hyperlink/url too ?

christo4ferris replied:
could you please clarify... you want a link to this README from where? (I am assuming that is what you're asking for here)

BiscuitBomber replied:
Sorry the suggestion wasn't clear. I meant to suggest: "as described in [Application ACLs](https://github.com/hyperledger/fabric/blob/master/docs/tech/application-ACL.md)"
 instead of "as described in **\*application-ACL.md***" [in line 33]

That is:  \[Application ACLs\] \(https://github.com/hyperledger/fabric/blob/master/docs/tech/application-ACL.md)

adecaro replied:
@BiscuitBomber @christo4ferris Fixed. Thanks for the feedback :)

BiscuitBomber commented on examples/chaincode/go/rbac/rbac.go:
```diff
@@ -0,0 +1,305 @@
+		return nil, errors.New("Failed getting metadata.")
+	}
+	if len(deployer) == 0 {
+		return nil, errors.New("Invalid admin certificate. Empty.")
```
"Empty.": Maybe "Missing caller metadata in admin certificate", "Deployer information is required in admin certificate", or so?

(trying to be a bit more informative in reporting the error.)

adecaro replied:
Hi @BiscuitBomber, what do you think of the following: "Missing metadata. Admin certificate expected in the metadata field."?

JonathanLevi replied:
@adecaro: this is marvelous! Thank you.

JonathanLevi replied:
@adecaro I think that this is marvelous! Thank you.

JonathanLevi replied:
@adecaro I think that this is a marvelous error message!

BiscuitBomber replied:
Oh, I see - I have learned something new from your proposed change already.

How about "An admin certificate must be provided in the metadata field."? Although I don't have a strong preference either way, so long as the message is instructive.

adecaro replied:
Deal :)

jeffgarratt commented:
@adecaro @srderson @binhn Want to hold this PR until it is reconciled with TCert with attrs example.

ghaskins commented:
Can this be closed and then reopened when its ready?

christo4ferris commented:
this needs some unit/function tests

wenjianqiao committed:
[FAB-17831] Use generic constant/var names in dataformat.go

Currently Version1x and Version20 are defined in dataformat.go.
They should be renamed to more generic names such as
PreviousFormat and CurrentFormat.

The format values will change only when the data format is changed
in ledger. If a Fabric version does not introduce a new data format,
CurrentFormat will remain the same as the latest format prior to
the Fabric version.

Signed-off-by: Wenjian Qiao <wenjianq@gmail.com>

cendhu commented:
@wenjianqiao Can you give a bit more context here? When we move to version 3, will we change the current version to 3.0 and previous version to 2.x?

manish-sethi commented:
> @wenjianqiao Can you give a bit more context here? When we move to version 3, will we change the current version to 3.0 and previous version to 2.x?

@cendhu - yes, that was the intent

cendhu commented on common/ledger/dataformat/dataformats.go:
```diff
@@ -9,16 +9,19 @@ package dataformat
-	// Version1x specifies the data format in version 1.x
-	Version1x = ""
+	// PreviousVersion specifies the data format in previous version (1.x for now)
+	PreviousVersion = ""
```
```suggestion
	LastMajorVersion = ""
```
This is because both 2.0 and 2.1 are called as two different versions. It would be good if we can be explicit.

manish-sethi replied:
These versions are not as such related to fabric versions. As you see the package name is `dataformat`, so they would change only when the data format is changed.

wenjianqiao replied:
@cendhu Thank you for your comments. As @manish-sethi  said, these versions are for the data formats in ledger databases (e.g, dataformat.CurrentVersion). They may or may not change when fabric version is upgraded. Currently, we don't have the concept for "major" data format version.

cendhu replied:
I see. Thanks. I think the confusion was due to the comment `CurrentVersion specifies the data format in the current version (2.0 for now)` and we have `CurrentVersion = "2.0"`. In that case, can we rename `CurrentVersion` to `CurrentDataVersion` and `PreviousVersion` to `PreviousDataVersion`.   Even some other name is fine but let's be more explicit to avoid hidden meaning. I think it is better to remove `2.0` as well but we cannot do that now, right? Some of the networks that have migrated to 2.0 would already have this format. Usually, `2.x` can be interpreted as 2 -- major version, x -- minor version. I assume we don't support major.minor version for the data format. When we change the data format in the future, we can use non-decimal number to avoid these confusions.

manish-sethi replied:
I am fine with the names `CurrentDataVersion` and `PreviousDataVersion`, if @wenjianqiao is fine. 
Yes, version is just a string for us and we cannot change `2.0`. The intention was to keep track when this format was introduced than anything else. So, in a way, that a binary is expected to work with the greatest data format only where  `data format version <= binary version`. @wenjianqiao - may be some of this description, you may like to add in the data format file.

wenjianqiao replied:
Talked offline and agreed to change the constants to CurrentFormat and PreviousFormat.

cendhu commented on common/ledger/dataformat/dataformats.go:
```diff
@@ -9,16 +9,19 @@ package dataformat
-	// Version20 specifies the data format in version 2.0
-	Version20 = "2.0"
+	// CurrentVersion specifies the data format in current version (2.0 for now)
+	CurrentVersion = "2.0"
```
```suggestion
	CurrentMajorVersion = "2.0"
```

cendhu reviewed (commented):
Thanks, @manish-sethi . Does my suggestion make sense? Only you know the clear context here and I leave it to you.

wenjianqiao commented:
Hit a CI flake. Log is attached to FAB-17451.
FAIL: TestWithStaticDeliverClientLeader (46.00s)
gossip_service_test.go:626: Failed to establish full channel membership. Only 0 out of 2 peers have full membership

wenjianqiao commented:
/ci-run

wenjianqiao commented:
/azp run

lindluni commented:
/azp run

cendhu commented on common/ledger/dataformat/dataformats.go:
```diff
@@ -9,16 +9,19 @@ package dataformat
+	PreviousVersionBytes []byte = nil
+
+	// CurrentVersionBytes specifies the []byte for CurrentVersion
+	CurrentVersionBytes = []byte(CurrentVersion)
```
Do we need these? Can't the consumer simply convert the version string to bytes when needed?

wenjianqiao replied:
The format is stored as []byte in a db. There was no format key in fabric v1.x, so we compare to "nil" for upgrade eligibility. However, we cannot convert the PreviousVersion string to nil bytes. CurrentVersionBytes can be converted from CurrentVersion string, but it is added to be same as PreviousVersionBytes. I don't have strong opinion to keep or remove CurrentVersionBytes.

cendhu replied:
A general suggestion is to avoid package-level variables (for Fabric). I am not getting into that. My question is more about why we need to define variables such as **XXXbytes** to just represent a byte version of **XXX** const. I am fine with this usage when it is used within a function as local variables but certainly not as exported variables (when both **XXX** and **XXXbytes** are exported). 

Can't we use 
```
if bytes.Equal(format, []byte(dataformat.PreviousVersion)) 
if bytes.Equal(format, []byte(dataformat.CurrentVersion))
```
 instead of 
```
if bytes.Equal(format, dataformat.PreviousVersionBytes)
if bytes.Equal(format, dataformat.CurrentVersionBytes)
```
?

Even if the stored version of the data is nil, can't we compare it against `[]byte(dataformat.PreviousVersion)` given that the `dataformat.PreviousVersion` is `""`? 

The following bytes.Equal would return true. 
```
var dbVersion1x []byte = nil
var previousVersion = ""
fmt.Println(bytes.Equal(dbVersion1x, []byte(previousVersion)))
fmt.Println(bytes.Equal(nil, []byte(previousVersion)))
fmt.Println(bytes.Equal(nil, []byte("")))
fmt.Println(bytes.Equal([]byte(nil), []byte("")))
```
You can play with this code here -- https://play.golang.org/p/WywdtwFTGvo

Let me know if I am missing anything.

wenjianqiao replied:
Thank you. Will remove XXXBytes.
